File: scripts/generate_proof.js
const fs = require('fs');
const crypto = require('crypto');
const { execSync } = require('child_process');

try {
  // 1) get commit hash
  const commit = execSync('git rev-parse --verify HEAD').toString().trim();
  if (!/^[0-9a-f]{40}$/.test(commit)) throw new Error('Invalid commit hash: ' + commit);

  // 2) load keys (must exist in repo root)
  if (!fs.existsSync('student_private.pem')) throw new Error('Missing student_private.pem');
  if (!fs.existsSync('instructor_public.pem')) throw new Error('Missing instructor_public.pem');
  const priv = fs.readFileSync('student_private.pem', 'utf8');
  const instrPub = fs.readFileSync('instructor_public.pem', 'utf8');

  // 3) sign commit hash (ASCII) using RSA-PSS SHA-256, max salt
  const signer = crypto.createSign('sha256');
  signer.update(commit, 'utf8');
  signer.end();
  const signature = signer.sign({
    key: priv,
    padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
    saltLength: crypto.constants.RSA_PSS_SALTLEN_MAX_SIGN
  });

  // 4) encrypt signature with instructor public key using OAEP-SHA256
  const encrypted = crypto.publicEncrypt({
    key: instrPub,
    padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
    oaepHash: 'sha256'
  }, signature);

  // 5) output
  console.log('Commit Hash:', commit);
  console.log('Encrypted Signature (base64):', encrypted.toString('base64'));
} catch (e) {
  console.error('ERROR:', e && (e.stack || e.message || e));
  process.exit(1);
}

File: scripts/generate_proof.py
#!/usr/bin/env python3
import subprocess
import base64
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

STUDENT_KEY = "student_private.pem"
INSTRUCTOR_KEY = "instructor_public.pem"


def get_commit_hash():
    out = subprocess.check_output(["git", "log", "-1", "--format=%H"]).strip()
    return out.decode('utf-8')


def load_private_key(path):
    with open(path, 'rb') as f:
        return serialization.load_pem_private_key(f.read(), password=None)


def load_public_key(path):
    with open(path, 'rb') as f:
        return serialization.load_pem_public_key(f.read())


def sign_message(message: str, private_key) -> bytes:
    return private_key.sign(
        message.encode('utf-8'),
        padding.PSS(
            mgf=padding.MGF1(hashes.SHA256()),
            salt_length=padding.PSS.MAX_LENGTH,
        ),
        hashes.SHA256(),
    )


def encrypt_with_public_key(data: bytes, public_key) -> bytes:
    return public_key.encrypt(
        data,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None,
        ),
    )


def main():
    commit = get_commit_hash()
    priv = load_private_key(STUDENT_KEY)
    sig = sign_message(commit, priv)
    pub = load_public_key(INSTRUCTOR_KEY)
    enc = encrypt_with_public_key(sig, pub)
    b64 = base64.b64encode(enc).decode('utf-8')
    print("Commit Hash:", commit)
    print("Encrypted Signature (base64):")
    print(b64)


if __name__ == '__main__':
    main()
